lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,477 @@
+            UpdateState(completedTask.Id, TaskStateEvent.CompletedTask);
+        }
+
+        /// <summary>
+        /// This method is called when receiving IFailedTask event during task submitting or running
+        /// Removes the task from running tasks if the task was running
+        /// Updates the task state to fail based on the error message in the failed task
+        /// </summary>
+        /// <param name="failedTask"></param>
+        internal void RecordFailedTaskDuringRunningOrSubmissionState(IFailedTask failedTask)
+        {
+            //// Remove the task from running tasks if it exists there
+            _runningTasks.Remove(failedTask.Id);
+            UpdateState(failedTask.Id, GetTaskErrorEvent(failedTask));
+        }
+
+        /// <summary>
+        /// This method is called when receiving IFailedTask event during system shutting down. 
+        /// If the task failed because it receives the close command from driver, update the task state to TaskClosedByDriver.
+        /// Task could fail by communication error or any other application or system error during this time, as long as it is not 
+        /// TaskFailedByEvaluatorFailure, update the task state based on the error received. 
+        /// </summary>
+        /// <param name="failedTask"></param>
+        internal void RecordFailedTaskDuringSystemShuttingDownState(IFailedTask failedTask)
+        {
+            var taskState = GetTaskState(failedTask.Id);
+            if (taskState == StateMachine.TaskState.TaskWaitingForClose)
+            {
+                UpdateState(failedTask.Id, TaskStateEvent.ClosedTask);
+            }
+            else if (taskState != StateMachine.TaskState.TaskFailedByEvaluatorFailure)
+            {
+                UpdateState(failedTask.Id, GetTaskErrorEvent(failedTask));
+            }
+        }
+
+        /// <summary>
+        /// This method is called when receiving an IFailedEvaluator event during TaskSubmitted, TaskRunning or system shutting down.
+        /// Removes the task from RunningTasks if the task associated with the FailedEvaluator is present and running. 
+        /// Sets the task state to TaskFailedByEvaluatorFailure 
+        /// </summary>
+        /// <param name="failedEvaluator"></param>
+        internal void RecordTaskFailWhenReceivingFailedEvaluator(IFailedEvaluator failedEvaluator)
+        {
+            if (failedEvaluator.FailedTask.IsPresent())
+            {
+                var taskId = failedEvaluator.FailedTask.Value.Id;
+                var taskState = GetTaskState(taskId);
+                if (taskState == StateMachine.TaskState.TaskRunning)
+                {
+                    RemoveRunningTask(taskId);
+                }
+
+                if (GetTaskState(taskId) == StateMachine.TaskState.TaskWaitingForClose)
+                {
+                    UpdateState(taskId, TaskStateEvent.ClosedTask);
+                }
+                else
```

Comment:
I did not understand your comment. What I am simply saying is that in your transition definitions, you can set the following transition -> if task is in TaskWaitingForClose state and receives TaskStateEvent.FailedTaskEvaluatorError go to ClosedTask state. This is if part of loop. I guess else part transitions are already well define sin TaskStateTransition machine....This way you do not need if-else loop.